academia/signals.py

```diff
@@ -1,13 +1,67 @@
+# Standard imports
+import requests
+
+# Django imports
+from django.conf import settings
 from django.core.cache import cache
-from django.db.models.signals import post_save
 from django.dispatch import receiver
-from .models import School
+from django.db.models.signals import post_save
+
+# Local imports
+from .models import School, Country, Faculty, Department, Degree
+
+def webhook(type, fid):
+
+    # Dispatch webhook
+    API_URLS = settings.RECIEVER_WEBHOOK_API_URLS
+    webhook_urls = API_URLS.split(' ')
 
-def clear_school_cache(sender, instance, **kwargs):
-    print("Clearing cache")
+    for webhook_url in webhook_urls:
+        
+        # Include ID in the request body
+        payload = {
+                    'fid': str(fid),
+                    'type': type
+                  }
+        response_hook = requests.post(webhook_url, data=payload)
+
+def invalidate_country_cache(sender, instance, **kwargs):
+
+    cache_key = f"countries_in_univast"
+    if cache.get(cache_key):
+        cache.delete(cache_key)
+
+def webhook_clear_school_cache(sender, instance, **kwargs):
+    # Clear cache for school
     country_code = instance.country.country_code
-    print(country_code)
-    cache.delete(f"schools_{country_code}")
-    print("Cache cleared")
+    cache_key = f"schools_{country_code}"
+    if cache.get(cache_key):
+        cache.delete(cache_key)
+
+    # Dispatch webhook
+    webhook("school", instance.id)
+ 
+def webhook_clear_faculty_cache(sender, instance, **kwargs):
+    school_code = instance.school.code
+    cache_key = f"faculties_{school_code}"
+    if cache.get(cache_key):
+        cache.delete(cache_key)
     
-post_save.connect(clear_school_cache, sender=School)
\ No newline at end of file
+    # Dispatch webhook
+    webhook("faculty", instance.id)
+
+def webhook_clear_department_cache(sender, instance, **kwargs):
+    # Clear Cache
+    school_code = instance.school.code
+    faculty_name = instance.faculty.name
+    cache_key = f"departments_{school_code}_{faculty_name}"
+    if cache.get(cache_key):
+        cache.delete(cache_key)
+        
+    # Dispatch webhook
+    webhook("department", instance.id)
+
+post_save.connect(invalidate_country_cache, sender=Country)
+post_save.connect(webhook_clear_school_cache, sender=School)
+post_save.connect(webhook_clear_faculty_cache, sender=Faculty)
```

Comment:
I thought you've set up the server to support Redis and celery? If yes, is there a reason why you used Django signals? 

Django signals are not inherently asynchronous. As a matter of fact, it is a "blocking code". When a signal is triggered, the receivers (codes that are called in response to the signal) are executed synchronously, meaning that each request must complete before the next one can begin. 

Another caveat to using signals is the potential impact it'll have on scalability. When using synchronous signal handling, the application will have to wait for all the receivers to complete whatever they're doing before continuing with other requests (just like the default per-thread connection to the database with Django - Code blocking). 

Do not misunderstand me, man. Django signals are a powerful set of utility box that have helped, and will continue to help us in so many ways. However, if we're looking to start reducing the latency of our APIs, or reducing work load from our server, then we will need to start using asynchronous tasks with Celery and Redis (or RabbitMQ).

Reply:
Ah, yes. I didn't realise until now that this is for univast. It's all good, man.

Reply:
> Ah, damnit. I didn't realise until now that this repository is univast's.
> 
> _Life of an engineer that decided to check his mail without putting on his glasses, smh._
> 
> It's all good, man. Cancel out the use of asynchronous tasks.

It's all good man. I understand. 
I wanted to get the webhook to be dispatched and working then implement celery and Redis to make it asynchronous.

Reply:
All good.